### PR TITLE
[Catch2] adapt recipe for Conan v2

### PR DESCRIPTION
**Pull request by fpelliccioni**
Specify library name and version:  **catch2/3.1.0**

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/catch2/3.x.x/conanfile.py [hunk 1]
```diff
@@ -1,9 +1,13 @@
 import os
 import functools
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.cmake import CMake, CMakeToolchain
```
Comment (uilianries) [inline recipes/catch2/3.x.x/conanfile.py:5]:
```suggestion
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
```

File: recipes/catch2/3.x.x/conanfile.py [hunk 2]
```diff
@@ -57,50 +61,68 @@ def _compilers_minimum_version(self):
```
Comment (uilianries) [inline recipes/catch2/3.x.x/conanfile.py:61]:
```suggestion

    def layout(self):
        cmake_layout(self, src_folder="src")

```

File: recipes/catch2/3.x.x/conanfile.py [hunk 3]
```diff
@@ -57,50 +61,68 @@ def _compilers_minimum_version(self):
...
+        tc.generate()
 
     def build(self):
         self._patch_sources()
-        cmake = self._configure_cmake()
+        # cmake = self._configure_cmake()
+        # cmake.build()
+
+        cmake = CMake(self)
+        cmake.configure(build_script_folder=os.path.join(self.source_folder, os.pardir))
```
Comment (uilianries) [inline recipes/catch2/3.x.x/conanfile.py:110]:
```suggestion
        cmake.configure()
```
with cmake_layout, it should work

File: recipes/catch2/3.x.x/conanfile.py [hunk 4]
```diff
@@ -55,52 +59,73 @@ def _compilers_minimum_version(self):
...
+    #     if self.options.default_reporter:
+    #         cmake.definitions["CATCH_CONFIG_DEFAULT_REPORTER"] = self._default_reporter_str
+    #     cmake.configure(build_folder=self._build_subfolder)
+    #     return cmake
 
     def _patch_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        for x in self.conan_data.get("patches", {}).get(self.version, []):
+            patch(**x)
```
Comment (fpelliccioni) [inline recipes/catch2/3.x.x/conanfile.py:94]:
```suggestion
            patch(self, **x)
```

File: recipes/catch2/3.x.x/conanfile.py [hunk 5]
```diff
@@ -1,9 +1,13 @@
 import os
 import functools
```
Comment (prince-chrismc) [inline recipes/catch2/3.x.x/conanfile.py:2]: Linter :)

```suggestion
```

File: recipes/catch2/3.x.x/conanfile.py [hunk 6]
```diff
@@ -1,9 +1,13 @@
...
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.build import check_min_cppstd
+from conan.tools.scm import Version
+from conan.tools.files import apply_conandata_patches, copy, get, patch, rmdir
```
Comment (prince-chrismc) [inline recipes/catch2/3.x.x/conanfile.py:8]: Linter

```suggestion
from conan.tools.files import apply_conandata_patches, copy, get, rmdir
```

File: recipes/catch2/3.x.x/conanfile.py [hunk 7]
```diff
@@ -23,24 +27,14 @@ class Catch2Conan(ConanFile):
...
-    @property
-    def _build_subfolder(self):
-        return "build_subfolder"
 
     @property
     def _default_reporter_str(self):
         return '"{}"'.format(str(self.options.default_reporter).strip('"'))
 
     def export_sources(self):
-        self.copy("CMakeLists.txt")
```
Comment (prince-chrismc) [inline recipes/catch2/3.x.x/conanfile.py:41]: Please delte this file form git 🙏

**Review by prince-chrismc — COMMENTED** [review]
For #12888 

There a few extra. You need to update the test packages.

- rename the current one to `test_v1_package/`
- make a new one with old name for the new generators

FMT or rapidjson are the best examples I am using

**Review by uilianries — COMMENTED** [review]
@fpelliccioni please, move those cpp files from test_v1_package to test_package. When we fully migrate to Conan v2 and drop Conan v1, test_v1_package could be removed and those files will be missing for test_package.

**Comment by uilianries** [comment]
@fpelliccioni please, consider https://github.com/fpelliccioni/conan-center-index/pull/2

**Comment by fpelliccioni** [comment]
> fpelliccioni#2

thanks, merged

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**